Issue #24 - Solution

akdsco opened the pull request:
Hello,

I've created the PopularCard component. I installed additionally two dependencies, which I believe we should use for testing as well. Happy to hear feedback on them.

One is `@testing-library/jest-dom` which seems to be used in every test example I found for writing test with `react-testing-library`. I ended up using it as well and I would love to hear some feedback on my tests (as this is not my strongest area of expertise).

Second is `check-prop-types` dependency, which helps in testing if the component receives all the props specified by 'prop-types`. I think it might be useful to have it for the whole project. I'd like to know your opinion and if you decide not to use, I'll be happy to remove it.

Commit:
create basic CourseCardPopular component
- write story and render it in storybook

Commit:
add css to course card
- add link to index.html to fetch google fonts
- Roboto font not rendering in Storybook
- change component html structure slightly

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
simplify css naming, add placeholder, add render logic
- add code which makes sure values below "0" won't render
- add PropTypes to component
- add defaultProps
- update stories, add story which generates component with "1" as value

Commit:
add tests for PopularCard component
- rename component to PopularCard
- install 'check-prop-types' and 'jest-dom' dependencies for testing
- write initial unit tests
- restructure storybook rendering and naming

VirginiaBalseiro reviewed (changes requested):
Good work! Some comments:
- You don't need to create three separate stories for the different variants of the card. Instead, handle this though the *knobs* (see Storybook docs).
- I don't see the styling properly. When I run Storybook, the fonts don't match the design.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Update PopularCard component
- install storybook knobs devDependency
- use knobs in stories
- fix some bugs if props are not supplied

Commit:
move CourseCard to its own folder

akdsco commented:
Hello,

Have a look now. 

The fonts should render, however, I am still struggling to render them properly in my local environment. It does render when I explicitly @import them in my CSS file, but if not, they don't. I believe it must be something in my local setup.

I've used knobs and it even made me update the component to catch some null values.

Let me know if that's good enough.

maxvlc commented on frontend/src/components/PopularCard/PopularCard.css:
```diff
@@ -0,0 +1,89 @@
+.course-apply {
+  font-size: 14px;
+  color: black;
+}
```
Add newline at the end...

akdsco replied:
not sure what you mean exactly, just a new empty line at the end of the file? the same question applies to comment below

maxvlc replied:
Yes, just a code styling preference. As you prefer...

VirginiaBalseiro replied:
This should be handled by Prettier though

akdsco replied:
Might be that back then.. as I've coded this long ago, I didn't yet have prettier set up somehow, I remember that since WebStorm 2020, they baked it into the IDE, happy to run it through prettier if you like, let me know.

VirginiaBalseiro replied:
For now just adding the line (you can even edit it here in GitHub! no need to go to your editor) is fine, and just keep it in mind for next time!

maxvlc commented on frontend/src/test/testUtils.js:
```diff
@@ -0,0 +1,6 @@
+export const checkProps = (Component, conformingProps) => {
+  const propError = checkPropTypes(Component.propTypes, conformingProps, 'prop', Component.name);
+  expect(propError).toBeUndefined();
+};
```
also here

Enzo-Parodi commented on frontend/src/components/PopularCard/PopularCard.jsx:
```diff
@@ -0,0 +1,70 @@
+      ? ""
+      : concepts.reduce((first, second) => first.concat(", ", second));
+  const studentsStat = `${
+    studentsEnrolled < 0 || studentsEnrolled === null ? 0 : studentsEnrolled
```
You can count on `studentsEnrolled` to be always present since it's a required prop. Also, you can count on the number being always 0 or positive (otherwise we would have problem somewhere else, which this component is not responsible for). I would suggest using the prop directly.

Enzo-Parodi commented on frontend/src/components/PopularCard/PopularCard.jsx:
```diff
@@ -0,0 +1,70 @@
+  const studentsStat = `${
+    studentsEnrolled < 0 || studentsEnrolled === null ? 0 : studentsEnrolled
+  } ${studentsEnrolled === 1 ? "Student" : "Students"}`;
+  const reviewsStat = `${reviews < 0 || reviews === null ? 0 : reviews} ${
```
Same here.

Enzo-Parodi reviewed (commented):
Great job!

Enzo-Parodi commented on frontend/src/components/PopularCard/PopularCard.test.jsx:
```diff
@@ -0,0 +1,101 @@
+    studentsEnrolled: 3,
+    reviews: 1,
+  };
+  checkProps(PopularCard, expectedProps);
```
This test checks that the right prop types throw no error? I am not sure if we need a library for that kind of test. Or is it intended for something else besides this?
